pkg/manifests/tls.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2019 The Cluster Monitoring Operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package manifests
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/klog"
+
+	"github.com/openshift/library-go/pkg/crypto"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/authentication/user"
+)
+
+// Taken from
+// https://github.com/openshift/library-go/blob/08c2fd1b452520da35ad210930ea9d100545589a/pkg/operator/certrotation/signer.go#L68-L86
+// without refresh time handling. We just take care of rotation if we reach 1/5 of the validity timespan before expiration.
+func needsNewCert(notBefore, notAfter time.Time, now func() time.Time) bool {
+	maxWait := notAfter.Sub(notBefore) / 5
+	latestTime := notAfter.Add(-maxWait)
+	return now().After(latestTime)
+}
+
+func getSelfSignedCA(s *v1.Secret) (*crypto.CA, error) {
+	crt, crtPresent := s.Data["ca.crt"]
+	key, keyPresent := s.Data["ca.key"]
+
+	if crtPresent && keyPresent {
+		ca, err := crypto.GetCAFromBytes(crt, key)
+		if err != nil {
+			klog.Warningf("creating a new CA due to error reading CA: %v", err)
+		} else if !needsNewCert(ca.Config.Certs[0].NotBefore, ca.Config.Certs[0].NotAfter, time.Now) {
+			return nil, nil
```

Comment:
I feel like this check rather belongs in the calling function, I had to jump there and back to find the 

```
// no need to rotate
```

comment in the other function. At the very least we should comment this function with what the `nil` return means.

Reply:
agreed, good catch. I will refactor for better readability.